contracts/vaults/common/SingleSidedLPVaultBase.sol

```diff
@@ -0,0 +1,529 @@
+    /// emergency exit
+    function _lockVault() internal {
+        StrategyVaultState memory state = VaultStorage.getStrategyVaultState();
+        // Set locked flag
+        state.flags = state.flags | FLAG_LOCKED;
+        VaultStorage.setStrategyVaultState(state);
+        emit VaultLocked();
+    }
+
+    /// @notice Unlocks the vault, called during restore vault.
+    function _unlockVault() internal {
+        StrategyVaultState memory state = VaultStorage.getStrategyVaultState();
+        // Remove locked flag
+        state.flags = state.flags & ~FLAG_LOCKED;
+        VaultStorage.setStrategyVaultState(state);
+        emit VaultUnlocked();
+    }
+
+    /// @notice Allows the emergency exit role to trigger an emergency exit on the vault.
+    /// In this situation, the `claimToExit` is withdrawn proportionally to the underlying
+    /// tokens and held on the vault. The vault is locked so that no entries, exits or
+    /// valuations of vaultShares can be performed.
+    /// @param claimToExit if this is set to zero, the entire pool claim is withdrawn
+    function emergencyExit(
+        uint256 claimToExit, bytes calldata /* data */
+    ) external override onlyRole(EMERGENCY_EXIT_ROLE) {
+        StrategyVaultState memory state = VaultStorage.getStrategyVaultState();
+        if (claimToExit == 0) claimToExit = state.totalPoolClaim;
+
+        // By setting min amounts to zero, we will accept whatever tokens come from the pool
+        // in a proportional exit. Front running will not have an effect since no trading will
+        // occur during a proportional exit.
+        _unstakeAndExitPool(claimToExit, new uint256[](NUM_TOKENS()), true);
+
+        state.totalPoolClaim = state.totalPoolClaim - claimToExit;
+        state.setStrategyVaultState();
+
+        emit EmergencyExit(claimToExit);
+        _lockVault();
+    }
+
+    /// @notice Restores withdrawn tokens from emergencyExit back into the vault proportionally.
+    /// Unlocks the vault after restoration so that normal functionality is restored.
+    /// @param minPoolClaim slippage limit to prevent front running
+    function restoreVault(
+        uint256 minPoolClaim, bytes calldata /* data */
+    ) external override whenLocked onlyNotionalOwner {
+        StrategyVaultState memory state = VaultStorage.getStrategyVaultState();
+
+        (IERC20[] memory tokens, /* */) = TOKENS();
+        uint256[] memory amounts = new uint256[](tokens.length);
+
+        // All balances held by the vault are assumed to be used to re-enter
+        // the pool. Since the vault has been locked no other users should have
+        // been able to enter the pool.
+        for (uint256 i; i < tokens.length; i++) {
+            if (address(tokens[i]) == address(POOL_TOKEN())) continue;
+            amounts[i] = TokenUtils.tokenBalance(address(tokens[i]));
+        }
+
+        // No trades are specified so this joins proportionally using the
```

Comment:
I think in practice this could be tricky. Not allowing trades could result in significant cost because we have to put all the tokens back in at the same time. Depending on our size vs. the size of the pool, that could be an issue.

Reply:
If we ever even need to cross this bridge we can run an upgrade to allow for some trading I guess.